src/routes/snippets.rs

```diff
@@ -77,6 +131,68 @@ pub fn create_snippet(
     create_snippet_impl(&**storage, &snippet, origin.path())
 }
 
+#[allow(clippy::too_many_arguments)]
+#[get("/snippets?<title>&<syntax>&<tag>&<marker>&<limit>")]
+pub fn list_snippets<'o, 'h>(
+    storage: State<Box<dyn Storage>>,
+    origin: &'o Origin,
+    title: Option<String>,
+    syntax: Option<String>,
+    tag: Option<String>,
+    limit: Option<PaginationLimit>,
+    marker: Option<String>,
+    _content_type: NegotiatedContentType,
+    _user: BearerAuth,
+) -> Result<WithHttpHeaders<'h, Output<Vec<Snippet>>>, ApiError> {
+    let mut criteria = ListSnippetsQuery {
+        title,
+        syntax,
+        tags: tag.map(|v| vec![v]),
+        ..Default::default()
+    };
+
+    // Fetch one more record in order to detect if there's a next page, and
+    // generate appropriate Link entry accordingly.
+    criteria.pagination.limit = limit.map(|v| v.0).unwrap_or(criteria.pagination.limit) + 1;
+    criteria.pagination.marker = marker;
+
+    let limit = criteria.pagination.limit - 1;
+    let mut snippets = storage.list(criteria.clone())?;
+    let mut prev_needed = false;
+    let next_marker = if snippets.len() > limit {
+        snippets.get(limit - 1).map(|snippet| snippet.id.to_owned())
+    } else {
+        None
+    };
+    let prev_marker = if criteria.pagination.marker.is_some() {
+        // In order to generate Link entry for previous page we have no choice
+        // but to issue the query one more time into opposite direction.
+        criteria.pagination.direction = Direction::Asc;
+        criteria.pagination.marker = Some(snippets[0].id.to_owned());
+        let prev_snippets = storage.list(criteria.clone())?;
+        prev_needed = prev_snippets.len() >= limit;
```

Comment:
I tried to mimic whatever we have in python implementation. Apparently we have a bug there. :D 

You can send and request here - `api.xsnippet.org/v1/snippets?marker=364333` and check that there's no `prev` link in the output.

I wonder, how should we generate prev link in this case? I mean, we can't just generate the first page "As Is" since it will contain overlapping snippets. Should we also parametrize `limit` part to return only non-overlapping snippets?  WDYT?

Reply:
I think return non-overlapping snippets would be ideal, but yeah, that requires fiddling with the limit value :(

Reply:
> I think return non-overlapping snippets would be ideal, but yeah, that requires fiddling with the limit value :(

Which makes a flow a bit less than ideal if someone follow `prev` and a `limit` appears, and then follow `next` and a `limit` from `prev` will apply which may be not the limit they started their initial query. :thinking: